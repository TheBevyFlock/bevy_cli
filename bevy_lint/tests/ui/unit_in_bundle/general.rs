#![feature(register_tool)]
#![register_tool(bevy)]
#![deny(bevy::unit_in_bundle)]
#![allow(dead_code)]

use bevy::prelude::*;

#[derive(Component, Clone)]
struct ComponentA;

#[derive(Component, Clone)]
struct ComponentB;

fn generic_param<B: Bundle>(_bundle: B) {}
fn impl_trait(_bundle: impl Bundle) {}

fn many_bounds<A: Clone, B: Bundle, C: Bundle + Clone>(_a: A, _b: B, _c: C) {}

struct Receiver(u8);

impl Receiver {
    fn associated_fn<B: Bundle>(_bundle: B) {}
    fn method<B: Bundle>(&mut self, _bundle: B) {}
}

struct BundleWrapper<B: Bundle>(B);

impl<B: Bundle> BundleWrapper<B> {
    fn associated_fn(_bundle: B) {}
    fn method(&self, _bundle: B) {}
}

pub fn test_generic_params() {
    let unit = ();
    let contains_unit = (ComponentA, ());

    generic_param(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    generic_param(
        //~v ERROR: created a `Bundle` containing a unit `()`
        unit,
    );

    generic_param((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));

    generic_param((
        //~v ERROR: created a `Bundle` containing a unit `()`
        contains_unit,
        ComponentB,
    ));

    fn returns_unit() {}

    generic_param(
        //~v ERROR: created a `Bundle` containing a unit `()`
        returns_unit(),
    );
}

pub fn test_impl_trait() {
    impl_trait(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    impl_trait((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));
}

pub fn test_many_bounds() {
    many_bounds(
        (),
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    many_bounds(
        (),
        (
            ComponentA,
            (
                //~v ERROR: created a `Bundle` containing a unit `()`
                (),
            ),
        ),
        (ComponentA, ComponentB),
    )
}

pub fn test_receiver() {
    Receiver::associated_fn(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    Receiver::associated_fn((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));

    let mut receiver = Receiver(0);

    receiver.method(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    receiver.method((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));

    Receiver::method(
        &mut receiver,
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );
}

pub fn test_bundle_wrapper() {
    BundleWrapper::associated_fn(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    let unit_wrapper = BundleWrapper(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    unit_wrapper.method(
        //~v ERROR: created a `Bundle` containing a unit `()`
        (),
    );

    let bundle_wrapper = BundleWrapper((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));

    bundle_wrapper.method((
        ComponentA,
        (
            //~v ERROR: created a `Bundle` containing a unit `()`
            (),
            ComponentB,
        ),
    ));
}
