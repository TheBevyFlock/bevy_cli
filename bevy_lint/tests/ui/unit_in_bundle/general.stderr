error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:39:9
   |
39 |         (),
   |         ^^
   |
   = note: units `()` are not `Component`s and will be skipped
note: the lint level is defined here
  --> tests/ui/unit_in_bundle/general.rs:3:9
   |
 3 | #![deny(bevy::unit_in_bundle)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:44:9
   |
44 |         unit,
   |         ^^^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:51:13
   |
51 |             (),
   |             ^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:58:9
   |
58 |         contains_unit,
   |         ^^^^^^^^^^^^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:66:9
   |
66 |         returns_unit(),
   |         ^^^^^^^^^^^^^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:73:9
   |
73 |         (),
   |         ^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:80:13
   |
80 |             (),
   |             ^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:90:9
   |
90 |         (),
   |         ^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
  --> tests/ui/unit_in_bundle/general.rs:92:9
   |
92 |         (),
   |         ^^
   |
   = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:101:17
    |
101 |                 (),
    |                 ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:111:9
    |
111 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:118:13
    |
118 |             (),
    |             ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:127:9
    |
127 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:134:13
    |
134 |             (),
    |             ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:142:9
    |
142 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:149:9
    |
149 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:154:9
    |
154 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:159:9
    |
159 |         (),
    |         ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:166:13
    |
166 |             (),
    |             ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: created a `Bundle` containing a unit `()`
   --> tests/ui/unit_in_bundle/general.rs:175:13
    |
175 |             (),
    |             ^^
    |
    = note: units `()` are not `Component`s and will be skipped

error: aborting due to 20 previous errors

